api/interceptor/observer.go

```diff
@@ -0,0 +1,269 @@
+// the server handling perspective.
+func (*Observer) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, stream connect.StreamingHandlerConn) error {
+		rm := getRequestMetadata(ctx)
+		rm.procedure = stream.Spec().Procedure
+		rm.protocol = stream.Peer().Protocol
+
+		wrapped := streamingHandlerConn{
+			base: stream,
+			rm:   rm,
+		}
+		if err := next(ctx, wrapped); err != nil {
+			rm.err = err
+			return err
+		}
+		return nil
+	}
+}
+
+// RequestMetadata represents the metadata gathered during the entire lifecycle
+// of an incoming API request.
+type RequestMetadata struct {
+	startAt          time.Time
+	finishedAt       time.Time
+	duration         time.Duration
+	procedure        string
+	requestURI       string
+	method           string
+	protocol         string
+	peerAddress      string
+	messagesReceived int
+	messagesSent     int
+	bytesReceived    int64
+	bytesSent        int64
+	httpStatusCode   int
+	err              error
+}
+
+// StartAt retrieves the timestamp indicating when the server received the
+// request.
+func (r *RequestMetadata) StartAt() time.Time {
+	return r.startAt
+}
+
+// FinishedAt retrieves the timestamp indicating when the server completed
+// processing the request.
+func (r *RequestMetadata) FinishedAt() time.Time {
+	return r.finishedAt
+}
+
+// Duration returns the duration of the request processing.
+func (r *RequestMetadata) Duration() time.Duration {
+	return r.duration
+}
+
+// Procedure returns the RPC name.
+func (r *RequestMetadata) Procedure() string {
+	return r.procedure
+}
+
+// RequestURI returns the HTTP request uri.
+func (r *RequestMetadata) RequestURI() string {
+	return r.requestURI
+}
+
+// Method returns the HTTP method.
+func (r *RequestMetadata) Method() string {
+	return r.method
+}
+
+// Protocol returns the protocol that was used (e.g. grpc, connect, http).
+func (r *RequestMetadata) Protocol() string {
+	return r.protocol
+}
+
+// PeerAddress returns the IP address of the requester.
+func (r *RequestMetadata) PeerAddress() string {
+	return r.peerAddress
+}
+
+// MessagesReceived returns the number of messages received
+// in a streaming request.
+func (r *RequestMetadata) MessagesReceived() int {
+	return r.messagesReceived
+}
+
+// MessagesSent returns the number of messages sent
+// in a streaming request.
+func (r *RequestMetadata) MessagesSent() int {
+	return r.messagesSent
+}
+
+// BytesReceived is the number of bytes we read via HTTP.
+func (r *RequestMetadata) BytesReceived() int64 {
+	return r.bytesReceived
+}
+
+// BytesSent is the number of bytes we sent via HTTP.
+func (r *RequestMetadata) BytesSent() int64 {
+	return r.bytesSent
+}
+
+// HTTPStatusCode returns the HTTP status code. This is expected to be 200 for
+// most gRPC and connectrpc requests.
+func (r *RequestMetadata) HTTPStatusCode() int {
+	return r.httpStatusCode
+}
```

Comment:
Should we also have equivalent `RPCCode()` or something similar that returns the `connect.Code` from the `err`, or `0`. Or performs translation / mapping something similar to [this](https://github.com/connectrpc/vanguard-go/blob/main/protocol_http.go#L64)?

Reply:
Yes I can return add a ConnectCode method in addition to `StatusCode()` which is the string version. The reason I didn't want to return a connect.Code is that there's no connect.Code for status okay unfortunately.